[MRG + 1] Directly compute polynomial features.

eamartin opened the pull request:
Polynomial features are computed by iterating over all combinations
of features. For each combination of features, the product of the
columns indexed by the combination is computed.

The fit method is now a no-op, and the transform method works with any
number of features (regardless of what fit was called with).

Both the memory usage and the runtime are favorable to #3512 and #3514.

I believe the total memory usage is twice the size of the output array
because the output array is stored both as the argument to hstack and
as the final output. This might be able to be reduced by pre-computing
the output size and writing directly to the output array (rather than
using the hstack call).

eamartin committed:
Directly compute polynomial features.

Polynomial features are computed by iterating over all combinations
of features. For each combination of features, the product of the
columns indexed by the combination is computed.

The fit method is now a no-op, and the transform method works with any
number of features (regardless of what fit was called with).

Both the memory usage and the runtime are favorable to #3512 and #3514.

I believe the total memory usage is twice the size of the output array
because the output array is stored both as the argument to hstack and
as the final output. This might be able to be reduced by pre-computing
the output size and writing directly to the output array (rather than
using the hstack call).

jnothman commented:
Nice, if it works! But I think it remains useful to store the matrix, for introspection's sake.

lightcatcher commented:
Here are the benchmarking results on my system.

```python
n_samples, n_features = 100000, 20
X = np.arange(-n_samples, n_samples * n_features - n_samples).reshape((n_samples, n_features))
p = PolynomialFeatures(degree=2)
````

PR | Time | Memory
---- | ------ | -----------
#3512 | 9.77s | 176.43MiB
#3514 | 2.37s | 743.86MiB
#4286  | 1.95s | 353.64MiB

I think the memory saved (and the speed advantage) makes up for not storing the ``powers_`` matrix. I also find the code easier to understand without the ``powers_`` matrix.

The data matrix is ~175MB, and my memory usage is twice the data matrix. I'm going to update my code now to avoid the extra copy of the data matrix (so #4286 should use only ~175MB on this benchmark).

eamartin committed:
PolynomialFeatures has no memory overhead.

PolynomialFeatures.transform now precomputes the size of the output
matrix. This allows the method to avoid having two copies of the
output data, and cuts the memory overhead to close to 0.

The size of the output matrix is computed by iterating over the
combinations generator an extra time. An alternative would be to
sum all of the necessary binomial terms, but iterating over the
combinations generator is (1) easier and (2) adds very little overhead
compared to the second iteration through the generator which involves
a lot of IO and multiplications.

lightcatcher commented:
After commit a0326e3, this PR uses 176.43MiB (the size of the output matrix) and takes the same 1.9s.

eamartin committed:
Maintain identical PolynomialFeatures semantics.

PolynomialFeatures now throws an error if you try to call transform
before calling fit or with a different number of features than with
which fit was called.

eamartin committed:
Add author.

jnothman commented:
Could you also benchmark with 1e6 data points (although I guess as this
grows kernel methods become more appropriate); using more features, degree
2; and using 20 features, higher degree?

On 24 February 2015 at 20:26, Coveralls <notifications@github.com> wrote:

>  [image: Coverage Status] <https://coveralls.io/builds/1984917>
>
> Coverage decreased (-0.0%) to 95.08% when pulling *7dd5aed
> <https://github.com/scikit-learn/scikit-learn/commit/7dd5aedc9287a6b9ee9edbea1694a8bfd057f4fb>
> on lightcatcher:fast_small_polynomial_features* into *f6af488
> <https://github.com/scikit-learn/scikit-learn/commit/f6af4881a5a66fb21688379b39f9304898a11bc0>
> on scikit-learn:master*.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/4286#issuecomment-75723893>
> .
>

lightcatcher commented:
I timed ``fit_transform`` for all benchmarks because this is the method users are most likely to use and because it best represents the total memory usage of the class.

(5e5, 20), degree=2

| PR  | Time | Memory |
| ------------ | ----------- | ------------------- |
| #3512   |  59.4s  | 880.26MiB |
|  #3514  | 13.0s |   3087.74MiB |
| #4286   | 10.8s  | 881.37MiB |

Notably, #3514 caused swapping on my machine and took 4+ minutes of wall time.

(500, 500), degree=2

| PR  | Time | Memory |
| ------------ | ----------- | ------------------- |
| #3512   |  >600s  | ???  |
|  #3514  |    95s   |  1910.75MiB |
| #4286   | 6.56s  | 476.25MiB |

(2000, 20), degree=4

| PR  | Time | Memory |
| ------------ | ----------- | ------------------- |
| #3512   |  10.6s  |  162.29MiB |
|  #3514  |    3.17s   |  635.08MiB |
| #4286   | 2.1s  | 162.17MiB |

eamartin committed:
Replace xrange with range

amueller commented:
sweet :)

jnothman commented on sklearn/preprocessing/data.py:
```diff
@@ -460,23 +455,22 @@ def __init__(self, degree=2, interaction_only=False, include_bias=True):
                                           self.interaction_only,
                                           self.include_bias)
+        self.n_input_features_ = n_features
+        self.n_output_features_ = sum(1 for _ in combinations)
```
I think it's worth providing `powers_` so the user has a description of each feature. Alternatively, `powers_` could be calculated dynamically as a `@property` to avoid the `fit` overhead.

You could calculate `n_output_features_` using stats.special.binom if you cared to. But I gather the current implementation is no big deal.

amueller commented:
Looks good to me, but you can't just remove the ``powers_`` attribute, users might rely on it (it was introduced before 0.15).
So you need to keep it if possible, event if it might not be used. The order of the output features is not changed, right?

Also, you should add an entry to ``doc/whatsnew.rst``.

jnothman commented on sklearn/preprocessing/data.py:
```diff
@@ -439,12 +440,6 @@ class PolynomialFeatures(BaseEstimator, TransformerMixin):
            [ 1,  2,  3,  6],
            [ 1,  4,  5, 20]])
 
-    Attributes
```
The attributes should still be documented

lightcatcher commented:
@amueller The order of the output features is not changed. I'm going to calculate ``powers_`` dynamically using ``@property``. 

#3514 warns that ``powers_`` is deprecated and will be removed from version 0.18. I'd like to deprecate ``powers_``, but can it still be removed in 0.18 or does it need to be a later version?

mblondel commented:
What I would really like to see is sparse matrix support.

amueller commented:
@mblondel is that not somewhat orthogonal to this PR?

amueller commented:
@lightcatcher Using ``@property`` sounds good, version 0.18 is fine :)

amueller commented:
Btw I don't have a strong opinion on keeping or deprecating ``powers_``. @jnothman seems to think it is helpful. not sure what others think.

amueller commented:
Btw @mblondel you can open an issue, there are currently quite a few people that pick up on feature requests and improvements :)

mblondel commented:
@amueller Sorry, I was just requesting a feature :) Yes, it's orthogonal if we use two different implementations but I think it may be possible to come up with an implementation that works on both dense and sparse formats.

eamartin committed:
Restore powers_ and document.

The powers_ field on PolynomialFeatures is now implemented as a
property and deprecated (as its no longer necessary).

PolynomialFeatures.transform uses np.empty while allocating the output
array rather than np.zeros because the rest of the method assigns values
to the array.

Documentation exists for n_input_features_ and n_output_features_. A
note in the "What's New" was also created.

eamartin committed:
Merge remote-tracking branch 'upstream/master' into fast_small_polynomial_features

Conflicts:
	doc/whats_new.rst

lightcatcher commented:
I've updated the documentation, added back ``powers_``, and modified "what's new". I also changed a call to ''np.zeros'' to ''np.empty''.

I'm ready for this to merge if any of you are.

Thanks to all for the help/feedback with making my first contribution to scikit-learn!

jnothman commented:
#3514 deprecates powers_ but adds a sparse equivalent. And no one reviewed
that design favourably!

On 25 February 2015 at 15:47, Eric Martin <notifications@github.com> wrote:

>  I've updated the documentation, added back powers_, and modified "what's
> new". I also changed a call to ''np.zeros'' to ''np.empty''.
>
> I'm ready for this to merge if any of you are.
>
> Thanks to all for the help/feedback with making my first contribution to
> scikit-learn!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/4286#issuecomment-75905024>
> .
>

lightcatcher commented:
I don't have a strong opinion about deprecating ``powers_``. Its no longer needed so it makes sense that it would eventually disappear, but it can stay indefinitely if preferred.

jnothman commented:
I think there needs to be *some* means of model introspection that doesn't require understanding the implementation.

lightcatcher commented:
Isn't the ``_combinations`` function good enough for model introspection? It seems like a more intuitive explanation of the data (especially if it was a list rather than a generator) than ``powers_`` to me.

jnothman commented:
It's fine, but it's equivalent to a sparse matrix, and there is already an
interface to return a matrix. I'm happy with a combinations_ attribute that
is a list, if we believe that's a substantial gain over powers_

On 25 February 2015 at 16:57, Eric Martin <notifications@github.com> wrote:

>  Isn't the _combinations function good enough for model introspection? It
> seems like a more intuitive explanation of the data (especially if it was a
> list rather than a generator) than powers_ to me.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/4286#issuecomment-75909924>
> .
>

lightcatcher commented:
Having
```python
@property
def combinations_(self):
    return list(self._combinations(...))
```
seems a little bit funky to me, but this does seem like the best interface if we require some variable describing the output. Providing the combinations seems superior to providing the matrix (sparse or not) because it is simpler to actually compute the polynomial terms given the list of combinations than it is given the powers_ matrix.

Does renaming ``_combinations`` to ``_combination_iter``, adding the ``combinations_`` property, and leaving ``powers_`` deprecated sound like a good resolution to you @jnothman ?

jnothman commented:
If we were designing the tool anew, I'd say `combinations_` is absolutely the right answer. Now it seems like a bit of a willy-nilly deprecation, and we may be better off just leaving `powers_`. I'm willing to go with the deprecation plan mostly because this class hasn't been around for long. I'd also be happy to hear another opinion.

lightcatcher commented:
@jnothman I added in the deprecation of ``powers_`` because it isn't used any more and because its what you did in #3514.

@mblondel I think a very similar algorithm can work for sparse matrices, and will be fast for CSC matrices. The only part which needs to be converted is
```python
for i, c in enumerate(combinations):
    XP[:, i] = X[:, c].prod(1)
```.
The column slicing is fast with CSC, I've figured out how to do the ``np.prod(axis=1)`` equivalent for CSC, and the return value can be COO and constructed efficiently using ``np.fromiter``. The only part that is a little sketchy is my prod implementation, which uses an internal Scipy call ``_minor_reduce``. Is it alright if Scikit-Learn depends on Scipy internals like this? I'm thinking writing a sparse prod reduction in ``utils.extmath`` and also submitting a patch to Scipy with the prod code, and then implementing sparse polynomial features.

jnothman commented:
You certainly can't use `_minor_reduce`, partially because it's private, and partially because it's only found in a very recent scipy (I know, I wrote it) beyond our minimum requirements.

lightcatcher commented:
So what are we waiting for to get this merged in? Are we waiting for some more opinions on deprecating ``powers_``, or just for to add in a ``combinations_`` property and the appropriate documentation?

amueller commented:
basically we are waiting for consensus on what would be a good way to expose the feature composition to users. I'm fine with the deprecation, but I don't have strong opinion.

amueller commented:
I'm sure @ogrisel has an opinion ;)

lightcatcher commented:
Any progress/discussion?

I'd like to see this merge as its a significant performance improvement in a part of the codebase that I actually use. Any chance we can merge the current version (which deprecates ``powers_`` and introduces no documented alternative) and discuss deprecation in the future? Alternatively, I could not deprecate ``powers_`` and then deprecation could still be discussed in the future.

mjbommar commented:
+1 to wanting this!  Tested this branch on a >1000 feature model and it's night-vs-day compared to master...no more r3.8xl's!

amueller commented:
Sorry for the delay. I have no strong opinion as I said before, and no-one else seems to find the time to comment :-/

glouppe commented:
CC: Since you are the original author @jakevdp, if you find time, I think your opinion is needed :)

amueller commented:
@glouppe do you have any opinion on the interface? We could just not deprecate and keep the old and the new interface and have a much better implementation.

jnothman commented:
I'm happy with keeping old and new; I don't really care much as long as
some attrib helps the user decode the construction of the features.

On 17 March 2015 at 06:57, Andreas Mueller <notifications@github.com> wrote:

>  @glouppe <https://github.com/glouppe> do you have any opinion on the
> interface? We could just not deprecate and keep the old and the new
> interface and have a much better implementation.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/4286#issuecomment-81904422>
> .
>

amueller commented:
Ok then let's remove the deprecation and merge.

amueller commented:
Do we want this in 0.16 @ogrisel ?

eamartin committed:
Remove deprecation of powers_.

lightcatcher commented:
Alright, I've removed the deprecation and the tests pass. Can I go ahead and merge?

amueller commented:
You still have my :+1:. @jnothman ok with green button?

jnothman commented on sklearn/preprocessing/data.py:
```diff
@@ -441,9 +442,13 @@ class PolynomialFeatures(BaseEstimator, TransformerMixin):
+    n_input_features_ : int
+        The total number of input features.
 
-    powers_ :
```
This should be reinstated.

jakevdp replied:
I agree. I have code that uses this attribute; it's not a huge deal to continue storing it.

jnothman commented:
Merged as ed47b5b. @ogrisel, please pull into 0.16 as an efficiency improvement.

jnothman commented:
Thank you @albahnsen for the initial report and patch.

amueller commented:
Thanks for the merge :) this is a pretty sweet improvement.

lightcatcher commented:
Glad to help. I'm probably going to write polynomial feature computation for sparse input sometime in the next 2 weeks. This will have to loop over all of the examples and perform different computation for each, so Cython will probably be worthwhile.

jnothman commented:
Although again, sparse matrix multiplication in log space is a
straightforward solution...

On 18 March 2015 at 15:12, Eric Martin <notifications@github.com> wrote:

>  Glad to help. I'm probably going to write polynomial feature computation
> for sparse input sometime in the next 2 weeks. This will have to loop over
> all of the examples and perform different computation for each, so Cython
> will probably be worthwhile.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/4286#issuecomment-82731686>
> .
>

albahnsen commented:
Hi guys, thanks for this improvement, I just test it with the database I used when I first found the issue and it works great. Thanks @jnothman for all the effort you put in this.

mblondel commented:
> Although again, sparse matrix multiplication in log space is a straightforward solution...

Sounds interesting. Could you explain a bit more?

On a related note, I've added feature combination support to FeatureHasher in #4316. I still need to add an example and fix a bug when using Python 3, though.

jakevdp commented:
Thanks – sorry I couldn't participate in review.
